README.txt
YouTube Downloader with GUI
* goal is to replicate the user experience of downloading youtube videos with clipconverter.cc,
  but allow user to queue downloads like with the Steam downloads page

[MODULES]
io
 * used to convert raw image data into bytes for the thumbnail
PIL (Image, ImageTk)
 * used to open bytes into an image, to resize the image, to turn the image into a PhotoImage object
pytube (YouTube)
 * used for all functions related to YouTube videos
threading
 * used to perform multi-threading (e.g. to allow downloads to be performed in the background)
time (*)
 * used to turn YouTube video length (seconds: int) into a string (H:M:S format)
tkinter (*, messagebox)
 * used for the GUI
tkinter.ttk (*)
 * used for the progress bar
urllib.request
 * used to access pages behind URLs

CHANGELOG.txt
20230429
YT-DL-GUI Version 0.13
* added a timer to the loading_search function to check how long each search takes
	* initially noticed that each search took around 20 seconds, which felt too slow; tried to check the duration of each major process such as retrieving date_label
	* found that retrieving the streams took the bulk of the time	
		* weirdly enough, retrieving the streams first made retrieving the video title and video length FAST (0.1s)
	* as of now, no way to optimize retrieving the streams without checking under the hood
* placed retrieving the streams at the top of the search process (right after initializing the youtube object) so that, to the user, everything loads at once

20230419
YT-DL-GUI Version 0.12
* implemented download queue, meaning the user can simply add streams to the queue and the streams download in the background
	* opted not to make an abstract queue_stream class that contains all the details needed
	* will search for a way to optimize the process in the future
* implemented a display for the download queue; separated widgets related to search and widgets related to download
* loading label is not unmapped anymore to prevent widgets from shuffling vertically
* fixed bug where toggle button doesn't reset back to pause when download is cancelled
YT-DL-GUI Version 0.11
* implemented label that shows "Loading..." while details and options are still loading
* handled exception related to HTTP Error 404 
YT-DL-GUI Version 0.10
* implemented string slicing on the video title if it's too long
	* string[0:END:-1] doesn't work; string[0:END][::-1] does (to get the index of whitespace within the char_per_line limit)
* implemented exception handling for filenames that are invalid or already taken by an existing file
	* when implementing download queue in the future, must also check if filename was already taken by an item in the queue
* implemented feature where if the user cancels the download, the incomplete download file is deleted
* moved the filename check above the configuration of buttons to avoid having to reset the buttons if the filename was already taken

20230418
YT-DL-GUI Version 0.09
* the search button now starts a thread of search_stream, just like how the download button starts a thread of download_stream
* removed the use of StringVar for some widgets; went with directly changing the ["text"] of widgets instead to use less variables
* for now, placed all features of functions related to search_stream such as display_image and display_stream into the search_stream functions
	* this is to avoid errors related to one thread loading before the other but lacking required information because the other function hasn't finished yet
* wrapped yt.length in float() for gmtime() to help avoid the error "int() argument must be a string, a bytes-like object or a real number, not 'NoneType'"
	* root of problem is sometimes yt.length is not retrieved at all
* handled specific exceptions encountered when retrieving details such as youtube title, views, length

20230417
YT-DL-GUI Version 0.08
* implemented pause/resume (toggle) and cancel download features
	* subject to change when download queue is implemented, but for now, the download button is disabled while downloading is in progress or paused
* added a label that will display the download progress in text (bytes downloaded vs filesize)
* the user will be warned if the download button was clicked but no stream was selected, and the download will not start
* organized widgets into frames; main_window will consist only of frames that are packed while frames consist of widgets that are arranged in a grid
* implemented a feature where the frames of widgets disappear when the url is invalid, along with an error message
YT-DL-GUI Version 0.07
* user can now enter their preferred file name for the download file
* switched from stream.download() to manually writing from the iterable stream
	* this was necessary to introduce pause/resume and cancel features
	* this makes the on_progress and on_complete functions unnecessary, but will keep them for now to use as reference

20230416
YT-DL-GUI Version 0.06
* decided to change from camelCase to snake_case for consistency with tests of components
* uploaded tests of components to go back to in the future

20230415
YT-DL-GUI Version 0.06
* removed feature to specify video resolution and input stream identification tag (itag)
* replaced the above feature with a feature that shows user the list of available streams, 
  and user can download a stream by choosing from the list of options then clicking the download button
	* itag of each stream is contained in a global variable: list[int] and these are IDEALLY sorted in the same order as the streams in the listbox
* rearranged widgets to show in the following order:
	* enter youtube video URL
	* show youtube video details
	* download details
	* list of stream options

20230414
YT-DL-GUI Version 0.05
* thumbnail is now placed in a frame which also includes other details such as:
	* title, length (H:M:S), channel name, view count (in youtube format), date published (not accurate with date shown in page though)
	* could not compute video age (i.e. published xx ago) because publish date is off by a few days and lacks hours-minutes-seconds
YT-DL-GUI Version 0.04
* implemented a feature which shows the thumbnail of the video that was searched
	* the function for displaying the thumbnail is placed in a separate non-daemon thread
	* thumbnail is resized from 640x480 to 240x180
* implemented a "Show Threads" button for debugging purposes; to show the number of active threads at any time

20230412
YT-DL-GUI Version 0.03
* download function was given its own thread to prevent the Tk window from not responding while the download is going on
* the thread was set to be a daemon thread to allow the download to be interrupted/cancelled by closing the Tk window

20230411
YT-DL-GUI Version 0.02

test_list_streams.py
# test displaying list of available streams of the YouTube object
# stream: a video or audio track of the YouTube object
from tkinter import *
from pytube import YouTube
from time import *

yt: YouTube
tag_list: list[int] = []

def display_streams(video_url):
    global yt, tag_list
    item: StringVar
    yt = YouTube(url = video_url)

    for i in yt.streams.filter():
        print(i)
    
    tag_list.clear()
    options_listbox.delete(0, END)
    for stream in yt.streams.filter():
        if stream.type == "video":
            item = stream.type + " - " + stream.resolution + str(stream.fps) + " - " + stream.subtype + " (" + convertBytes(stream.filesize) + ")"
            if stream.is_adaptive:
                item += " (no audio)"
        elif stream.type == "audio":
            item = stream.type + " - " + stream.abr + " - " + stream.subtype + " (" + convertBytes(stream.filesize) + ")"
        options_listbox.insert(END, item)
        tag_list.append(stream.itag)

    options_frame.pack()

def choose_stream():
    global yt
    print(tag_list[options_listbox.curselection()[0]])
    # stream = yt.streams.get_by_itag(tag_list[options_listbox.curselection()[0]])
    # stream.download()

def convertBytes(bytes: int) -> str:
    if bytes < 1024:
        return "{i} B".format(bytes)
    elif bytes < (1024*1024):
        return "{:.2f} KB".format(bytes/1024)
    elif bytes < (1024*1024*1024):
        return "{:.2f} MB".format(bytes/1024/1024)
    else:
        return "{:.2f} GB".format(bytes/1024/1024/1024) # assuming video sizes don't get into terabyte territory
    
mainWindow = Tk()

url_entrybox = Entry(mainWindow)
display_button = Button(mainWindow, text = "Display Options", command = lambda: display_streams(url_entrybox.get()))

options_frame = Frame(mainWindow)
options_label = Label(options_frame, text = "Options:")
filter_button = Button(options_frame, text = "Choose Stream", command = choose_stream)
options_listbox = Listbox(options_frame, width = 50)

options_label.grid(row = 0, column = 0)
filter_button.grid(row = 0, column = 1)
options_listbox.grid(row = 1, column = 0, columnspan = 2)

url_entrybox.pack()
display_button.pack()

mainWindow.mainloop()


test_pause_cancel.py
# helping hand
# from https://stackoverflow.com/questions/63533960/how-to-cancel-and-pause-a-download-using-pytube#:~:text=You%20can%20try%20pytube.,the%20download%20as%20you%20wish.
import tkinter as tk
import threading
from pytube import YouTube, request

is_paused = False
is_cancelled = False

def download_video(url):
    global is_paused, is_cancelled
    download_button['state'] ='disabled'
    pause_button['state'] ='normal'
    cancel_button['state'] ='normal'
    try: 
        progress['text'] = 'Connecting ...'
        yt = YouTube(url)
        stream = yt.streams.first() # specify the desired stream from the YouTube object
        filesize = stream.filesize  # get the video size
        print(filesize)
        with open('sample.mp4', 'wb') as f: # since 'wb' then if a download was cancelled and the same stream was downloaded again then it will overwrite the old file, not append to it
            is_paused = is_cancelled = False
            print(stream.url)
            stream = request.stream(stream.url) # get an iterable stream which is read in chunks
                # stream.url is the url specific to the stream specified in line 18; streams of diff resolutions don't share the same url
            downloaded = 0
            while True:
                if is_cancelled:
                    progress['text'] = 'Download cancelled'
                    break
                if is_paused:
                    continue
                chunk = next(stream, None) # get next chunk of video
                if chunk: # if chunk exists; is not None
                    f.write(chunk)
                    downloaded += len(chunk)
                    progress['text'] = f'Downloaded {downloaded} / {filesize}'
                else:
                    # no more data
                    progress['text'] = 'Download completed'
                    break
        print('done')
    except Exception as e:
        print(e)
    download_button['state'] ='normal'
    pause_button['state'] ='disabled'
    cancel_button['state'] ='disabled'

def start_download():
    threading.Thread(target=download_video, args=(url_entry.get(),), daemon=True).start()

def toggle_download():
    global is_paused
    is_paused = not is_paused
    pause_button['text'] = 'Resume' if is_paused else 'Pause'

def cancel_download():
    global is_cancelled
    is_cancelled = True

root = tk.Tk()
root.title("YouTube Downloader")

tk.Label(root, text='URL:').grid(row=0, column=0, sticky='e')
url_entry = tk.Entry(root, width=60)
url_entry.grid(row=0, column=1, columnspan=2, padx=10, pady=10)

download_button = tk.Button(root, text='Download', width=20, command=start_download)
download_button.grid(row=1, column=2, sticky='e', padx=10, pady=(0,10))

pause_button = tk.Button(root, text='Pause', width=10, command=toggle_download, state='disabled')
pause_button.grid(row=2, column=0)

progress = tk.Label(root)
progress.grid(row=2, column=1, sticky='w')

cancel_button = tk.Button(root, text='Cancel', width=10, command=cancel_download, state='disabled')
cancel_button.grid(row=2, column=2, sticky='e')

root.mainloop()

test_details_frame.py
# test displaying basic details about YouTube video
from tkinter import *
from pytube import YouTube
import urllib.request
import io
from PIL import Image, ImageTk
import threading
from time import *

photoimage_holder: PhotoImage

def display_result(video_url):
    yt = YouTube(url = video_url)

    display_image_thread = threading.Thread(target = display_image, args = (yt.thumbnail_url, ))
    display_image_thread.start()

    if yt.length < 3600:   
        video_length = "Length: " + strftime("%M:%S", gmtime(yt.length)) # gmtime converts the int yt.length into a tuple to be used for strftime
    else:        
        video_length = "Length: " + strftime("%H:%M:%S", gmtime(yt.length))
    
    if yt.views < 1000:
        video_views = f"{yt.views} views"
    elif yt.views < 1000*10:
        video_views = f"{yt.views/1000: .1f}K views"
    elif yt.views < 1000*1000:
        video_views = f"{int(yt.views/1000)}K views"
    elif yt.views < 1000*1000*10:
        video_views = f"{yt.views/1000/1000: .1f}M views"
    else:
        video_views = f"{int(yt.views/1000/1000)} M views"

    title_label.configure(text = yt.title)
    time_label.configure(text = video_length)
    channel_label.configure(text = yt.author)
    views_label.configure(text = video_views)
    date_label.configure(text = "Published on: " + yt.publish_date.strftime("%b %m, %Y"))
    details_frame.pack()

def display_image(thumbnail_url):
    global image_holder
    # print(thumbnail_url)
    with urllib.request.urlopen(thumbnail_url) as image_url:
        thumbnail_data = image_url.read()
        thumbnail_bytes = io.BytesIO(thumbnail_data)
        with Image.open(thumbnail_bytes) as thumbnail_image:
            resize_image = thumbnail_image.resize((240, 180))
            thumbnail_photoimage = ImageTk.PhotoImage(resize_image)
            thumbnail_box.configure(image = thumbnail_photoimage)
    # keep a reference to PhotoImage object so that it appears properly
    image_holder = thumbnail_photoimage 

mainWindow = Tk()

url_entrybox = Entry(mainWindow)
display_button = Button(mainWindow, text = "Display Thumbnail", command = lambda: display_result(url_entrybox.get()))

details_frame = Frame(mainWindow)
thumbnail_box = Label(details_frame)
title_label = Label(details_frame)
time_label = Label(details_frame)
channel_label = Label(details_frame)
views_label = Label(details_frame)
date_label = Label(details_frame)

thumbnail_box.grid(row = 0, column = 0, rowspan = 5)
title_label.grid(row = 0, column = 1)
time_label.grid(row = 1, column = 1)
channel_label.grid(row = 2, column = 1)
views_label.grid(row = 3, column = 1)
date_label.grid(row = 4, column = 1)

url_entrybox.pack()
display_button.pack()

mainWindow.mainloop()


downloader.py
from tkinter import *
from tkinter import messagebox
from tkinter.ttk import *
from pytube import YouTube, request, exceptions
from time import *
import threading
import urllib.request
import urllib.error
import io
import os
from PIL import Image, ImageTk

yt: YouTube = None
video_title: str
photoimage_holder: PhotoImage = None
tag_list: list[int] = []
is_paused: bool = False
is_cancelled: bool = False
search_is_loading: bool = False
download_queue: list = []

def search_stream():
    global yt, video_title, photoimage_holder, tag_list, search_is_loading
    search_button["state"] = DISABLED     
    youtube_url: str = url_entrybox.get()
    try:
        yt = YouTube(youtube_url) 
    except exceptions.RegexMatchError:
        messagebox.showerror(title = "ERROR", message = "Invalid URL")
        search_button["state"] = NORMAL
        return
    search_is_loading = True
    threading.Thread(target = loading_search, args = (), daemon = True).start()
    """for some reason, loading the streams ahead avoids errors in retrieving data such as yt.title, yt.length, etc."""
    # try:
    #     yt.check_availability()
    # except Exception as e:
    #     messagebox.showerror(title = "ERROR", message = "The video is not available (" + e + ")")
    #     search_button["state"] = NORMAL
    #     return        
    # yt.bypass_age_gate()
    count: int = 0
    while True:
        try:
            streams = yt.streams
            break
        except KeyError as e:
            count += 1
            print(f"Failed to load streams, retrying {count} times")
    # LOAD SEARCH RESULTS
    try:
        # load the thumbnail into the thumbnail_box
        with urllib.request.urlopen(yt.thumbnail_url) as image_url:
            thumbnail_data = image_url.read()
            thumbnail_bytes = io.BytesIO(thumbnail_data)
            with Image.open(thumbnail_bytes) as thumbnail_image:
                effective_image = thumbnail_image.resize((240, 180))
                thumbnail_photoimage = ImageTk.PhotoImage(effective_image)
                thumbnail_box["image"] = thumbnail_photoimage
        # keep a reference to PhotoImage object so that it appears properly
        photoimage_holder = thumbnail_photoimage
    except urllib.error.HTTPError:
        messagebox.showerror(title = "ERROR", message = "Invalid URL (HTTP Error 404: Not Found)")
        search_button["state"] = NORMAL
        search_is_loading = False
        return
    # for some reason, pytube sometimes cannot properly retrieve the details of the video, requiring us to reload the YouTube object yt until the details are retrieved properly
    while True:
        try:
            video_title = yt.title
            break
        except Exception as e:
            yt = YouTube(youtube_url)
            continue
    # slice the video title to shorten the width on display
    char_per_line: int = 40
    if len(video_title) > char_per_line:
        reversed_first_line = video_title[0:char_per_line][::-1] # for some reason, slicing via indexing [0:char_per_line:-1] doesn't work
        whitespace_index = char_per_line - reversed_first_line.find(" ")
        character_limit = whitespace_index + char_per_line - 3 # display 2 lines maximum + "..." at the end
        if len(video_title) > character_limit:
            title_label["text"] = video_title[0:whitespace_index] + "\n" + video_title[whitespace_index:character_limit] + "..."
        else:
            title_label["text"] = video_title[0:whitespace_index] + "\n" + video_title[whitespace_index:]
    else:
        title_label["text"] = video_title    
    try:
        length: int = yt.length
        if length < 3600:   
            time_label["text"] = "Length: " + strftime("%M:%S", gmtime(length)) # gmtime converts the int yt.length into a tuple to be used for strftime
        else:        
            time_label["text"] = "Length: " + strftime("%H:%M:%S", gmtime(length)) # as of 2023-Apr-19, no way to display 1-9 instead of 01-09 for the hour
    except TypeError:
        time_label["text"] = "Length: Unavailable"
    try:
        views: int = yt.views
        if views < 1000:
            views_label["text"] = f"{views} views"
        elif views < 1000*10:
            views_label["text"] = f"{views/1000: .1f}K views"
        elif views < 1000*1000:
            views_label["text"] = f"{int(views/1000)}K views"
        elif views < 1000*1000*10:
            views_label["text"] = f"{views/1000/1000: .1f}M views"
        else:
            views_label["text"] = f"{int(views/1000/1000)} M views"
    except TypeError:
        views_label["text"] = "Views: Unavailable"
    channel_label["text"] = yt.author
    date_label["text"] = "Published on: " + yt.publish_date.strftime("%b %m, %Y")
    # DISPLAY STREAM OPTIONS
    item: str
    filename_entrybox.delete(0, END)
    filename_entrybox.insert(0, video_title)
    tag_list.clear()
    options_listbox.delete(0, END)
    for stream in streams:
        # print(stream) # for debugging; to check if tag_list and listbox match
        if stream.type == "video":
            item = stream.type + " - " + stream.resolution + str(stream.fps) + " - " + stream.subtype + " (" + convert_bytes(stream.filesize) + ")"
            if stream.is_adaptive:
                item += " (no audio)"
        elif stream.type == "audio":
            item = stream.type + " - " + stream.abr + " - " + stream.subtype + " (" + convert_bytes(stream.filesize) + ")"
        options_listbox.insert(END, item) # sometimes results in error: "RuntimeError: main thread is not in main loop"
        tag_list.append(stream.itag)

    search_is_loading = False
    search_button["state"] = NORMAL
    if not details_frame.winfo_ismapped():
        details_frame.pack()
    if not options_frame.winfo_ismapped():
        options_frame.pack()

def loading_search():
    global search_is_loading
    count: float = 0
    while search_is_loading:
        count += 1
        match count % 4:
            case 1:
                loading_label["text"] = "Loading"
            case 2:
                loading_label["text"] = "Loading."
            case 3:
                loading_label["text"] = "Loading.."
            case _:
                loading_label["text"] = "Loading..."
        sleep(1)
    loading_label["text"] = " "
    print(f"Finished in {count} seconds")
    return

def start_search():
    threading.Thread(target = search_stream, args = ()).start()

def advanced_options():
    print("You pressed the advanced options button.")
    # TODO post-processing options with FFMPEG in mind

def download_stream():
    global is_paused, is_cancelled, download_queue  
    if not download_frame.winfo_ismapped():
        download_frame.grid(row = 0, column = 1)
    # stream = yt.streams.get_by_itag(tag_list[options_listbox.curselection()[0]])
    # filename: str = filename_entrybox.get() + "." + stream.subtype
    # if os.path.exists(filename): # TODO when implementing download queue, must also check if filename has already been taken
    #     messagebox.showerror(title = "ERROR", message = "File name is already taken.")
    #     return
    # download_button["state"] = DISABLED
    pause_button["state"] = NORMAL
    cancel_button["state"] = NORMAL
    while len(download_queue) > 0:
        queue_listbox.delete(0)
        details: str = download_queue[0][0]
        filename: str = download_queue[0][1]
        filesize: int = download_queue[0][2]
        thumbnail: PhotoImage = download_queue[0][3]
        title: str = download_queue[0][4]
        stream_url: str = download_queue[0][5]
        current_downloading["text"] = "Now downloading: " + filename
        current_thumbnail["image"] = thumbnail
        current_details["text"] = title + "\n" + details
        converted_filesize: str = convert_bytes(filesize)
        with open(filename, "wb") as download_file:
            is_paused = False
            is_cancelled = False
            pause_button["text"] = "Pause"
            stream = request.stream(stream_url) # turn the stream into an iterable; pytube's default chunk size is 9MB
            bytes_downloaded: int = 0
            while True:
                if is_cancelled:
                    break
                if is_paused:
                    continue
                chunk = next(stream, None)  
                if chunk is not None:
                    download_file.write(chunk)
                    bytes_downloaded += len(chunk)
                    download_progress_bar["value"] = (bytes_downloaded / filesize) * 100
                    progress_label["text"] = convert_bytes(bytes_downloaded) + " / " + converted_filesize
                    main_window.update_idletasks()
                else:
                    # messagebox.showinfo(title = "Information", message = "Download complete.")
                    break
        if is_cancelled:
            # messagebox.showinfo(title = "Information", message = "Download was cancelled.")
            os.remove(filename)
        download_queue.pop(0)
    # download_button["state"] = NORMAL
    pause_button["state"] = DISABLED
    pause_button["text"] = "Pause"
    cancel_button["state"] = DISABLED
    download_progress_bar["value"] = 0
    progress_label["text"] = ""
    messagebox.showinfo(title = "Information", message = "Download queue is empty.")
    # download_frame.pack_forget()

def add_to_queue():
    # if len(options_listbox.curselection()) == 1:
    #     threading.Thread(target = download_stream, daemon = True).start()
    # else:
    #     messagebox.showinfo(title = "Information", message = "Choose a stream to download.")
    global yt, video_title, photoimage_holder
    if len(options_listbox.curselection()) == 0:
        messagebox.showerror(title = "ERROR", message = "Choose an option to download.")
        return
    stream = yt.streams.get_by_itag(tag_list[options_listbox.curselection()[0]])
    # print(tag_list[options_listbox.curselection()[0]])
    stream_filename: str = filename_entrybox.get() + "." + stream.subtype
    for char in stream_filename:
        if char in "\/:*?\"<>|":
            messagebox.showerror(title = "ERROR", message = "File name can't contain any of the following characters:\n\ / : * ? \" < > |")
            return
    if os.path.exists(stream_filename): # TODO when implementing download queue, must also check if filename has already been taken
        messagebox.showerror(title = "ERROR", message = "File name is already taken.")
        return
    stream_details: str = options_listbox.get(options_listbox.curselection())
    stream_filesize: int = stream.filesize
    stream_photoimage: PhotoImage = photoimage_holder
    stream_title: str = video_title
    stream_url: str = stream.url

    download_queue.append([stream_details,
                           stream_filename, 
                           stream_filesize,
                           stream_photoimage, 
                           stream_title,
                           stream_url])
    queue_details: str = stream_filename + " (" + convert_bytes(stream_filesize) + ")" + "\n" + stream_title + "\n" + stream_details
    queue_listbox.insert(END, queue_details)
    if len(download_queue) == 1:
        threading.Thread(target = download_stream, daemon = True).start()

def toggle_download():
    global is_paused
    is_paused = not is_paused
    if is_paused:
        pause_button["text"] = "Resume"
    else:
        pause_button["text"] = "Pause"

def cancel_download():
    global is_cancelled
    is_cancelled = True

# UTILITY FUNCTIONS
# convert bytes to KB/MB/GB for better readability
def convert_bytes(bytes: int) -> str:
    if bytes < 1024:
        return "{i} B".format(bytes)
    elif bytes < (1024*1024):
        return "{:.2f} KB".format(bytes/1024)
    elif bytes < (1024*1024*1024):
        return "{:.2f} MB".format(bytes/1024/1024)
    else:
        return "{:.2f} GB".format(bytes/1024/1024/1024)
# show threads for debugging 
def show_threads():
    print(threading.active_count())
    print(threading.enumerate())

main_window = Tk()

initial_frame = Frame(main_window)
download_frame = Frame(main_window)

search_frame = Frame(initial_frame)
details_frame = Frame(initial_frame)
options_frame = Frame(initial_frame)

initial_frame.grid(row = 0, column = 0)
search_frame.pack()

download_frame.grid(row = 0, column = 1)
download_frame.grid_forget()

# SEARCH WIDGETS
url_label = Label(search_frame, text = "Enter URL of YouTube Video:", width = 50, justify = "left")
url_entrybox = Entry(search_frame, width = 50)
search_button = Button(search_frame, text = "Search", command = start_search)
loading_label = Label(search_frame)
url_label.grid(row = 0, column = 0)
url_entrybox.grid(row = 1, column = 0)
search_button.grid(row = 1, column = 1)
loading_label.grid(row = 2, columnspan = 2)
# show threads for debugging
threads_button = Button(search_frame, text = "Threads", command = show_threads)
threads_button.grid(row = 0, column = 1)

# SEARCH VIDEO DETAILS WIDGETS
thumbnail_box = Label(details_frame)
title_label = Label(details_frame)
time_label = Label(details_frame)
channel_label = Label(details_frame)
views_label = Label(details_frame)
date_label = Label(details_frame)
thumbnail_box.grid(row = 0, column = 0, rowspan = 5)
title_label.grid(row = 0, column = 1)
time_label.grid(row = 1, column = 1)
channel_label.grid(row = 2, column = 1)
views_label.grid(row = 3, column = 1)
date_label.grid(row = 4, column = 1)

# STREAM OPTIONS WIDGETS
filename_label = Label(options_frame, text = "Enter preferred file name:", width = 50, justify = "left")
filename_entrybox = Entry(options_frame, width = 50)
download_button = Button(options_frame, text = "Download", command = add_to_queue)
options_label = Label(options_frame, text = "Options:")
advanced_options_button = Button(options_frame, text = "Advanced Options", command = advanced_options)
options_listbox = Listbox(options_frame, width = 50)
filename_label.grid(row = 0, column = 0)
filename_entrybox.grid(row = 1, column = 0)
download_button.grid(row = 1, column = 1)
options_label.grid(row = 2, column = 0)
advanced_options_button.grid(row = 2, column = 1)
options_listbox.grid(row = 3, column = 0, columnspan = 2)

# DOWNLOAD STREAM WIDGETS
current_downloading = Label(download_frame, text = "Now downloading: N/A")
current_thumbnail = Label(download_frame)
current_details = Label(download_frame)
download_progress_bar = Progressbar(download_frame, orient = HORIZONTAL, length = 200)
progress_label = Label(download_frame)
pause_button = Button(download_frame, text = "Pause", state = DISABLED, command = toggle_download)
cancel_button = Button(download_frame, text = "Cancel", state = DISABLED, command = cancel_download)
queue_label = Label(download_frame, text = "UP NEXT:", width = 50, justify = "left")
queue_listbox = Listbox(download_frame, width = 50)

current_downloading.grid(row = 0, column = 0, columnspan = 2)
current_thumbnail.grid(row = 1, column = 0, columnspan = 2)
current_details.grid(row = 2, column = 0, columnspan = 2)
download_progress_bar.grid(row = 3, column = 0, columnspan = 2)
progress_label.grid(row = 4, column = 0, columnspan = 2)
pause_button.grid(row = 5, column = 0)
cancel_button.grid(row = 5, column = 1)
queue_label.grid(row = 6, column = 0, columnspan = 2)
queue_listbox.grid(row = 7, column = 0, columnspan = 2)

main_window.mainloop()